### PR TITLE
Fixes to OpenAPI schema for bulk API operations

### PR DESCRIPTION
**Pull request by glennmatthews**
# Closes: #1619 
# What's Changed

Because Nautobot's bulk REST API operations (PUT, PATCH, DELETE) are bespoke implementations that are not part of the DRF standard operations set, the OpenAPI schema that drf-spectacular was generating for these operations was incomplete/incorrect in a number of ways, including:

- Request schema for bulk PUT/PATCH was showing as a single object representation, instead of what should be a list of object dictionaries with mandatory `id` field
- Request schema for bulk DELETE was showing no request body, instead of what should be a list of `{id: <uuid>}` dictionaries.
- Response schema for bulk PUT/PATCH was showing as a single object representation, instead of what should be a list

This PR includes the following fixes:

- Upgrade to latest drf-spectacular release 0.23.1
- Enable drf-spectacular `COMPONENT_SPLIT_PATCH` setting, allowing for distinct request schemas for PATCH versus POST/PUT
- Enable drf-spectacular `COMPONENT_SPLIT_REQUEST` setting, allowing for distinct schemas for request vs response
- The folllowing additions/changes to our custom `NautobotAutoSchema` subclass of drf-spectacular's `AutoSchema`:
  - ~Override ~internal `_get_request_body()`~ `get_operation()` method to allow for a request body for bulk-DELETE. drf-spectacular's base implementation here is hard-coded to assume that DELETE requests never have a request body, so we needed to override this to allow our exception to this case.
    - [x] ~TODO: investigate whether it would be possible/cleaner to inject this logic at a higher level without overriding an internal method, even if it means we need to write more lines of code~
  - Extend our overridden `get_request_serializer()` method implementation to handle bulk actions and PATCH actions specifically.
    - Extend our custom `get_writable_class()` method to be capable of generating `BulkWritable*Serializer` dynamic classes as well as the existing `Writable*Serializer` dynamic classes.
  - Override `get_response_serializers()` to recognize that bulk PUT/PATCH return a list of objects rather than a single object
    - Because bulk-PUT/PATCH return a bare list of objects (`[{...}, {...}, ...]`) instead of a standard REST API paginated-style list (`{"results": [{...}, {...}, ...], ...}`) and this list is not filterable like a GET endpoint, I also had to override `get_filter_backends()` and the internal `_get_paginator()` method to have both make an exception for bulk operations. (`_get_paginator` is called in a number of places fairly deep inside the guts of the AutoSchema class, so I don't think it'll be possible to avoid overriding it even though it's a private method)
      - I've added a couple of `2.0 TODO` comments for us to revisit this aspect of the REST API design to be more self-consistent.
  - Override `resolve_serializer()` to recognize that in our custom bulk-PATCH requests, the `id` field is always required (by default drf-spectactular assumes that there are no required fields in any PATCH request)
- Explicitly decorate the `BulkDestroyModelMixin.bulk_destroy` view to represent its request body as a `BulkOperationSerializer`
  - Change `BulkOperationSerializer` to always use a UUID as the `id` field (rather than a generic string), as this is more accurate for 99% of our models.
  - For the `Group` model, which as a Django built-in still uses integer PKs, add `BulkOperationIntegerIDSerializer` and use it where needed.
- Because we now have `BulkWritable*Serializer` as well as `Writable*Serializer`, we need to explicitly name the `Status` choices for *all* Status-containing model in `settings.py` to prevent drf-spectacular from complaining about collisions between enums.

Example bulk PUT/PATCH/DELETE schemas, showing list inputs and outputs and correct marking of required fields:

Screen Shot 2022-09-09 at 10 52 54 AM

Screen Shot 2022-09-09 at 10 53 19 AM

Screen Shot 2022-09-09 at 10 53 34 AM

Bulk PUT/PATCH/DELETE for Groups, with special-case integer PK:

Screen Shot 2022-09-09 at 10 54 01 AM

Screen Shot 2022-09-09 at 10 54 10 AM

Screen Shot 2022-09-09 at 10 54 19 AM

# TODO

- [x] Explanation of Change(s)
- [x] Attached Screenshots, Payload Example
- n/a Unit, Integration Tests
- n/a Documentation Updates (when adding/changing features)
- n/a Example Plugin Updates (when adding/changing features)
- n/a Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Review by bryanculver — APPROVED** [review]
Spot checked a few endpoints between generated schemas and actual function and looks good.

File: nautobot/core/api/schema.py [hunk 1]
```diff
@@ -145,6 +234,18 @@ def get_serializer_ref_name(self, serializer):
...
+        """
+        Re-add required `id` field on bulk_partial_update action.
+
+        drf-spectacular clears the `required` list for any partial serializers in its `_map_basic_serializer()`,
+        but Nautobot bulk partial updates require the `id` field to be specified for each object to update.
+        """
+        component = super().resolve_serializer(serializer, direction)
+        if self.is_bulk_action and self.is_partial_action and direction == "request":
+            component.schema["required"] = ["id"]
+        return component
```
Comment (jathanism) [inline nautobot/core/api/schema.py:247]: I hate that we have to do all of this and if there's any reasonable expectation that we can undo this in v2, can we please add "v2 TODO" for these?

File: nautobot/core/api/schema.py [hunk 2]
```diff
@@ -107,25 +174,47 @@ def get_writable_class(self, serializer):
...
+                            field.required = True
+                        new_fields[name] = field
+                    return new_fields
 
-        writable_class = self.writable_serializers[type(serializer)]
+                properties["get_fields"] = get_fields
+
+            self.writable_serializers[writable_name] = type(writable_name, (type(serializer),), properties)
+
+        writable_class = self.writable_serializers[writable_name]
```
Comment (jathanism) [inline nautobot/core/api/schema.py:217]: I hate that we have to do all of this and if there's any reasonable expectation that we can undo this in v2, can we please add "v2 TODO" for these?

File: nautobot/core/api/schema.py [hunk 3]
```diff
@@ -36,6 +37,49 @@ class NautobotAutoSchema(AutoSchema):
...
+
+            operation["requestBody"] = {
+                "content": {
+                    media_type: build_media_type_object(build_array_type(component.ref))
+                    for media_type in self.map_parsers()
+                },
+                "required": True,
+            }
+
+        return operation
```
Comment (jathanism) [inline nautobot/core/api/schema.py:81]: Another ripe opportunity for a "v2 TODO" I hope? Such as if we can align around normalized responses? Or is the lack of filtering/pagination for these bulk responses always going to be an issue here? (Such as how GET will always be paginated, but all the others won't?)

**Review by jathanism — COMMENTED** [review]
All in all this is pretty concise. My only concerns are around whether or not we can move away from this in v2?

File: nautobot/core/api/schema.py (hunk 1 above)
Comment (glennmatthews) [reply nautobot/core/api/schema.py:247]: I think bulk-PATCH will always require `id` as a field, as how else to identify the specific instances being patched? Unless we want to get rid of these bulk operations entirely in v2.

File: nautobot/core/api/schema.py (hunk 3 above)
Comment (glennmatthews) [reply nautobot/core/api/schema.py:81]: This part is specifically for bulk-DELETE; I think we could possibly push this upstream to get drf-spectacular to remove its code that hard-codes all DELETE request bodies to null.

File: nautobot/core/api/schema.py [hunk 4]
```diff
@@ -36,6 +37,49 @@ class NautobotAutoSchema(AutoSchema):
...
+        """Nautobot's custom bulk operations, even though they return a list of records, are NOT paginated."""
+        if self.is_bulk_action:
+            return None
+        return super()._get_paginator()
+
+    def get_filter_backends(self):
+        """Nautobot's custom bulk operations, even though they return a list of records, are NOT filterable."""
+        if self.is_bulk_action:
+            return []
+        return super().get_filter_backends()
```
Comment (glennmatthews) [inline nautobot/core/api/schema.py:60]: Regarding v2 - I don't think it makes sense for bulk-PATCH and bulk-PUT to support filter parameters or to provide true paginated responses (how would you retrieve "page 2" of a PATCH response?). I'd like to push an issue up to drf-spectacular to make `get_paginator` into a non-private method though.

File: nautobot/core/api/schema.py (hunk 2 above)
Comment (glennmatthews) [reply nautobot/core/api/schema.py:217]: I'm definitely open to ideas on how we could make the bulk and non-bulk serializers more similar. I don't see an obvious path to doing so though.

File: nautobot/core/api/schema.py (hunk 1 above)
Comment (jathanism) [reply nautobot/core/api/schema.py:247]: I suppose what I meant is if there's a way we can get away with it without requiring so much high touch. But perhaps now that this is solved it's less of a concern.

File: nautobot/core/api/schema.py (hunk 3 above)
Comment (jathanism) [reply nautobot/core/api/schema.py:81]: > This part is specifically for bulk-DELETE; I think we could possibly push this upstream to get drf-spectacular to remove its code that hard-codes all DELETE request bodies to null.

I like this idea.

**Review by jathanism — APPROVED** [review]

**Comment by glennmatthews** [comment]
Looks like drf-spectacular 0.24 was just released - I'd like to evaluate it and see if it adds anything worth including in this PR.

File: nautobot/core/api/schema.py (hunk 3 above)
Comment (glennmatthews) [reply nautobot/core/api/schema.py:81]: Opened https://github.com/tfranzel/drf-spectacular/issues/809.

File: nautobot/core/settings.py [hunk 5]
```diff
@@ -221,11 +221,24 @@
         "PowerPortTypeChoices": "nautobot.dcim.choices.PowerPortTypeChoices",
         "RackTypeChoices": "nautobot.dcim.choices.RackTypeChoices",
         "RelationshipTypeChoices": "nautobot.extras.choices.RelationshipTypeChoices",
-        # Because Interface and VMInterface, and Site and Location, have the same default statuses, we get the error:
+        # Each of these StatusModels has bulk and non-bulk serializers, with the same status options, so we'd get error:
```
Comment (itdependsnetworks) [inline nautobot/core/settings.py:224]: get `an` error?

File: nautobot/core/settings.py [hunk 6]
```diff
@@ -221,11 +221,24 @@
...
         #   enum naming encountered a non-optimally resolvable collision for fields named "status"
```
Comment (glennmatthews) [inline nautobot/core/settings.py:225]:
```suggestion
        # Each of these StatusModels has bulk and non-bulk serializers, with the same status options,
        # which confounds drf-spectacular's automatic naming of enums, resulting in the below warning:
        #   enum naming encountered a non-optimally resolvable collision for fields named "status"
        # By explicitly naming the enums ourselves we avoid this warning.
```

**Pull request merged**